core/src/main/java/com/linecorp/armeria/client/endpoint/ResourceFileEndpointGroup.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Properties;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ * A resource file backed {@link EndpointGroup}. The list of {@link Endpoint}s are loaded from the resource file
+ * in a current class loader.
+ */
+public final class ResourceFileEndpointGroup implements EndpointGroup {
+    /**
+     * Creates a new {@link EndpointGroup} instance that loads the host names (or IP address) and the port
+     * numbers of the {@link Endpoint} from the {@code resourceName} resource file. The resource file must
+     * contain at least one property whose name starts with {@code endpointKeyPrefix}:
+     *
+     * <pre>{@code
+     * example.hosts.0=example1.com:36462
+     * example.hosts.1=example2.com:36462
+     * example.hosts.2=example3.com:36462
+     * }</pre>
+     *
+     * @param resourceName the resource file name that the list of {@link Endpoint}s loaded from
+     * @throws IllegalArgumentException if failed to load any hosts from the specified resource file
+     */
+    public static ResourceFileEndpointGroup of(ClassLoader classLoader, String resourceName,
+                                               String endpointKeyPrefix, int defaultPort) {
+        checkArgument(defaultPort > 0, "defaultPort: %s (expected: > 0)", defaultPort);
```

Comment:
nit: How about adding the upper bound like we do in `DnsAddressEndpointGroup` or `Endpoint`?